### PR TITLE
vimPlugins: add ddc.vim and related plugins

### PR DESCRIPTION
**Pull request by euank**
## Description of changes

This adds the most commonly used set of ddc.vim plugins, or at least the set I use as part of my completion setup.

You can read more about ddc.vim over here: https://github.com/Shougo/ddc.vim

I've manually tested that these plugins work with my vim setup, and so far I've been maintaining an overlay for them. I'm hoping to be able to upstream them so others can benefit, and to simplify my own setup.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PerchunPak** [comment]
Those plugins depend on each other, could you please specify it similar to [here](https://github.com/NixOS/nixpkgs/blob/bd4f00ac7fa6b641777b5a7fb08911c456b4ed01/pkgs/applications/editors/vim/plugins/overrides.nix#L197-L199)?

File: pkgs/applications/editors/vim/plugins/overrides.nix
Comment (PerchunPak) [inline pkgs/applications/editors/vim/plugins/overrides.nix:558]: This plugin also depends on pum, so this PR is blocked by #293288, right?

Comment (euank) [reply pkgs/applications/editors/vim/plugins/overrides.nix:558]: Yup, I was about to comment to that effect!

I also still need to update my own vim config to point at this branch and verify all is well.

**Comment by euank** [comment]
I verified the rebased-and-updated plugins in this PR still work with my vim setup!

Other than adding the `pum` dependency, I'm happy with this.

File: pkgs/applications/editors/vim/plugins/overrides.nix
Comment (euank) [reply pkgs/applications/editors/vim/plugins/overrides.nix:687]: Done, I've added `pum-vim` here as well!

**Comment by euank** [comment]
Now that the `pum-vim` PR merged, I've updated `ddc-ui-pum` to have that dependency; as far as I know, this is good to go!

**Review by PerchunPak — APPROVED** [review]
LGTM! Thanks!

CC @teto

**Comment by teto** [comment]
shouldn't that depend on a `deno` runtime somewhow ?

**Comment by euank** [comment]
They do indirectly!

`denops-vim` depends on `deno` [here](https://github.com/NixOS/nixpkgs/blob/f000d7063413ad275ec262c037f62b258d836eb3/pkgs/applications/editors/vim/plugins/overrides.nix#L669-L674), `ddc-vim` depends on `denops-vim`, and all of these depend on `ddc-vim`.

I could explicitly add the `denops` dependency for all of them as well, but I believe as-is they'll always pull in deno via denops.

**Comment by teto** [comment]
my bad, was too lazy to check it out. Patching denops-vim (vimPlugins in general is tough). Is there a way to setup the path towards deno without patching the plugin, via configuration for instance, like `vim.g.denops_runtime = ...`. 
It's tangential to this PR that I approve, just curious.

**Review by teto — APPROVED** [review]

**Comment by euank** [comment]
There's indeed the [`g:denops#deno`](https://github.com/vim-denops/denops.vim/blob/45b8a1c45fe06e95538fc701ee7d6e3cc5335659/doc/denops.txt#L105C2-L105C15) setting, so people can override it if they want that way.

The current approach feels nice to me, since changing that external config's default value via adding an extra autoload file or such would, well, change the observable default value.

**Comment by teto** [comment]
thanks. Good to know. There is no clear stance on how to package the vim plugins. I would like to avoid patching as much as possible because it's bad practice (makes upstream debugging harder) but alternatives are not great for now.
I am thinking about having the neovim wrapper collect configurations from the plugins and autoadd it to its config ?!
You can read about this here https://github.com/NixOS/nixpkgs/issues/172538. 
Waiting for CI before merge

**Review by natsukium — APPROVED** [review]

**Pull request merged**
